app.go

```diff
@@ -1,248 +1,226 @@
+
+	if signError != nil {
+		log.Warn().Msgf("Couldnt sign signidice_part_2, reason: %s", signError.Error())
+		return
+	}
+
+	trx, packedTx, err := GetSigndiceTransaction(api, event.Sender, app.BlockChain.CasinoAccountName, event.RequestID,
+		signature)
+
+	if err != nil {
+		log.Warn().Msgf("couldn't form transaction, reason: %s", err.Error())
+		return
+	}
+
+	log.Debug().Msgf("%+v", trx)
+
+	result, sendError := api.PushTransaction(packedTx)
+	if sendError != nil {
+		log.Warn().Msg("Failed to send transaction, reason: " + sendError.Error())
+		return
+	}
+	log.Debug().Msg("Successfully signed and sent txn, id: " + result.TransactionID)
 }
 
-func (app *App) RunEventListener(parentContext context.Context, wg *sync.WaitGroup) {
-
-    go func(parentContext context.Context) {
-        defer wg.Done()
-        events := make(chan *broker.EventMessage)
-
-        listener := broker.NewEventListener(app.Broker.Url, events)
-        ctx, cancel := context.WithCancel(context.Background())
-
-        if err := listener.ListenAndServe(ctx); err != nil {
-            log.Panic().Msg(err.Error())
-        }
-
-        defer cancel()
-
-        offsetPath := app.Broker.TopicOffsetPath
-        offset := readOffset(offsetPath)
-        topicID := app.Broker.TopicID
-
-        log.Debug().Msgf("Subscribing to event type %+v with an offset of %+v", topicID, offset)
-        _, err := listener.Subscribe(topicID, offset)
-
-        if err != nil {
-            log.Error().Msg("Failed to subscribe")
-            return
-        }
-
-        for {
-           select {
-           case <-parentContext.Done():
-               log.Debug().Msg("Terminating event listener")
-               _, err := listener.Unsubscribe(topicID)
-               if err != nil {
-                   log.Warn().Msg("Failed to unsubscribe")
-               } else {
-                   log.Debug().Msg("Event listener successfully terminated")
-               }
-               return
-           case eventMessage, ok := <-events:
-               if !ok {
-                   log.Info().Msg("Failed to read events")
-                   break
-               }
-               if len(eventMessage.Events) == 0 {
-                   log.Debug().Msg("Gotta event message with no events")
-                   break
-               }
-               log.Debug().Msgf("Processing %+v events", len(eventMessage.Events))
-               for _, event := range eventMessage.Events {
-                   go app.processEvent(event)
-               }
-               offset = eventMessage.Events[len(eventMessage.Events) - 1].Offset + 1
-               writeOffset(offsetPath, offset)
-           }
-        }
-    }(parentContext)
+func (app *App) RunEventProcessor(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case eventMessage, ok := <-app.EventMessages:
+			if !ok {
+				log.Warn().Msg("Failed to read events")
```

Comment:
so here it's too big trouble

Reply:
it happens regularly, when you shut down the server for example

Reply:
Really strange, that's very bad if we can't detect real error and error caused by shutdown. And why shut down cause error...
@Garik-, do you have some advice here  ?